### ISSUE TITLE
Taking feature requests!

### ISSUE THREAD

**Issue by Joshuaalbert** [Joshuaalbert/jaxns#210]
What would you like to see in JAXNS? What would make your problem solving easier? 

1. Do you want better model representation (the probabilistic programming framework side)?
2. Do you want to see more examples combining ML and Bayesian computation?
3. Do you want better use of gradients (currently experimental)?
4. Do you want better default parameters?
5. Do you want better distributed solving (compute cluster problem solving)?
6. Do you want better solving super hard problems full of degeneracies?
7. Do you want to use normalising flows to incrementally repartition the problem so that it uses fewer likelihood evals, at the cost of overhead compute?
8. Do you want to see more special priors? Which ones?
9. Do you want more standardised benchmarks to assess accuracy in more problem settings?
11. Do you want better evidence-EM (currently experimental)?
12. Do you want better global likelihood maximisation (currently experimental)?
13. Something else?

**Comment by Joshuaalbert** [comment]
@avivajpeyi any thoughts?

**Comment by tomkimpson** [comment]
Hi @Joshuaalbert. Great work on `jaxns` 👏 ! 

One thing that I think would be helpful is some kind of checkpointing, e.g. https://dynesty.readthedocs.io/en/latest/quickstart.html#checkpointing

**Comment by Joshuaalbert** [comment]
Jacobs actually already has the ability to pause and resume but it's not documented. I will make it a main feature.On Jun 2, 2025 10:02, Tom Kimpson ***@***.***> wrote:tomkimpson left a comment (Joshuaalbert/jaxns#210)
Hi @Joshuaalbert. Great work on jaxns 👏 !
One thing that I think would be helpful is some kind of checkpointing, e.g. https://dynesty.readthedocs.io/en/latest/quickstart.html#checkpointing

**Comment by tavin** [comment]
I've faced some memory issues, so a nice "feature" for me would be improved memory efficiency 😄 

In particular, using multiple jax devices (cpus), the memory usage spikes by an order of magnitude somewhere near the end of the procedure (running + to_results()).

This often leads to an OOM kill. I think resolving this would not be the same thing as the proposed feature to resample points with low weights down to a smaller number of points.

**Comment by Joshuaalbert** [comment]
Hi @tavin I'll work on memory efficiency! To get the final log_L contour would `results.log_L_samples.max()` work? Interesting you're using the register, not many people use it. Since it's an internet component, it could change from under your feet. Maybe tell me what you're using it for and I can make sure you still have those features in version 3 which I'm aiming to deliver in the next month or so.

**Comment by Joshuaalbert** [comment]
@tavin  The memory usage spike could be several things. If you have a simple test case that exhibits it that would be helpful.

**Comment by Joshuaalbert** [comment]
> I've also observed a big deviation from the true distribution in the last of these samples

How do you mean? Since the number of live points diminishes from `n` to 1 for the final point, the 'weight' of each sample will grow as it performs the final quadrature for the evidence. These also represent the posterior sample weights. So we're effectively saying that the final points 'cover' larger swathes of posterior space in a Lebesgue sense.

**Comment by mvsoom** [comment]
I haven't used this package in a while, but I'm still keeping an eye on it :)
Would love to have similar feedback like Dynesty's:
```
iter: 2195 | bound: 135 | nc: 13 | ncall: 23838 | eff(%):  9.208 | loglstar:   -inf < 4653.113 <    inf | logz: 4631.879 +/-  0.437 | dlogz:  3.773 >  1.000  
```
After a while of looking at these, you can really "feel" how the run is going.
So if we could have a debug switch to print these (jax.debug.print) while exploring the inference process, that would be a major asset for me personally.

**Comment by Joshuaalbert** [comment]
We actually already have a verbose option in the nested sampler using JAX debug print callback. You're right it's helpful for seeing how things are going.On Aug 25, 2025 12:36, Marnix Van Soom ***@***.***> wrote:mvsoom left a comment (Joshuaalbert/jaxns#210)
I haven't used this package in a while, but I'm still keeping an eye on it :)
Would love to have similar feedback like Dynesty's:
iter: 2195 | bound: 135 | nc: 13 | ncall: 23838 | eff(%):  9.208 | loglstar:   -inf < 4653.113 <    inf | logz: 4631.879 +/-  0.437 | dlogz:  3.773 >  1.000  

After a while of looking at these, you can really "feel" how the run is going.
So if we could have a debug switch to print these (jax.debug.print) while exploring the inference process, that would be a major asset for me personally.

**Comment by mvsoom** [comment]
Alright, my bad, didn't know about that 👍

**Comment by pierrotlamontagne** [comment]
**Periodic/circular parameter support (e.g. orbital phase, angles)**

I ran into a concrete case where this would help: fitting exoplanet radial-velocity orbits, the time-of-conjunction epoch is naturally periodic (`phase in [0,1)`, `tc = t_ref - phase*P`). Two other samplers I use for the same class of problem already handle this natively.

`ultranest` has a native `wrapped_params=` flag: a per-parameter boolean telling its slice sampler to treat that unit-cube coordinate as periodic, so proposals wrap around 0/1 instead of stopping at the edge. `dynesty` has the analogous `periodic=[...]`/`reflective=[...]` parameter-index kwargs on its samplers. For a gradient-based sampler, `numpyro`/NUTS can be made wall-free with a `VonMises` prior plus `numpyro.infer.reparam.CircularReparam`, giving an unconstrained circular latent with no hard boundary at all.

For JAXNS, the corresponding `Prior` can only be a plain bounded `Uniform(0,1)` (checked against 2.6.9's `framework/prior.py` and both `uni_slice_sampler.py`/`multi_slice_sampler.py`, neither has any periodic/circular handling in the proposal step).

The practical failure mode: when the true solution's phase happens to land close to 0 or 1 (just a matter of where the reference epoch falls, not a property of the physics), the posterior mass straddles a hard prior-volume edge with no wraparound. That likely degrades exploration/ESS right at the boundary even though the underlying posterior is a perfectly ordinary unimodal lobe.

Something like a `periodic=True` flag on `Prior`, or a dedicated construction that wraps proposal steps around the unit interval for flagged dimensions, would remove this failure mode for any periodic/angular parameter (orbital phase, longitude, angles more generally), not just my RV use case.

Happy to share a minimal reproducer if useful.

**Comment by Joshuaalbert** [comment]
@pierrotlamontagne That's a great suggestion. I'll add it in. As a temporary solution, you have two options:
1. Use two variables `X~Normal(0,1)`, `Y~Normal(0,1)` and let `angle=arctan2(Y,X)`.
2. Or, use `angle_unwrapped ~ Uniform[-N*2*np.pi, N*2*np.pi]` for N >> 1 (like N=10 should be fine), and then take `angle = (angle_unwrapped + 2 * np.pi) % (2 * np.pi)` which wraps to `[0, 2pi)`.

**Comment by Joshuaalbert** [comment]
@pierrotlamontagne #275 will implement periodic parameters.
